blockprod/src/detail/builder.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use chainstate::ChainstateHandle;
+use common::{
+    chain::{Block, ChainConfig},
+    primitives::{BlockHeight, Id},
+    time_getter::TimeGetter,
+};
+use futures::FutureExt;
+use logging::log;
+use mempool::{MempoolEvent, MempoolHandle};
+use tokio::sync::mpsc;
+
+use crate::BlockProductionError;
+
+use super::block_maker::{BlockMaker, BlockMakerControlCommand};
+
+/// Master in the master/slave model with BlockMaker. Every time there's a new tip in chainstate or mempool,
+/// the perpetual block builder constructs a new instance of BlockMaker that keeps trying to create a block.
+pub struct PerpetualBlockBuilder {
```

Comment:
I'd suggest renaming this to something like `BlockProductionController`. Though I like the use of Perpetual here, it is confusing to have both Builder and Maker which are synonyms. Also it requires extensive comments with master/slave to describe the intent.